objectstorage/object_storage.go

```diff
@@ -927,14 +932,22 @@ func (objectStorage *ObjectStorage) flush() {
 
 	objectStorage.flushMutex.Unlock()
 
+	// shut down the executor to execute all release tasks
+	objectStorage.releaseExecutor.Shutdown(timedexecutor.IgnorePendingTimeouts)
+
 	// force release the collected objects
 	for j := 0; j < i; j++ {
 		if consumers := atomic.AddInt32(&(cachedObjects[j].consumers), -1); consumers == 0 {
-			objectStorage.options.batchedWriterInstance.Enqueue(cachedObjects[j])
+			cachedObjects[j].evict()
 		}
 	}
 
 	objectStorage.cachedObjectsEmpty.Wait()
+
+	if !shutdown {
+		// create a new release executor because the other was shut down to flush all objects
+		objectStorage.releaseExecutor = timedexecutor.New(objectStorage.options.releaseExecutorWorkerCount)
```

Comment:
Overwriting the releaseExecutor outside the scope of any mutex will most probably lead to race conditions but adding another mutex just for the rare occasion of where we flush also seems like anoverkill - maybe we can use atomic.LoadPointer whenever we access the releaseExecutor and atmon.StorePointer to set it?

Reply:
ok I will change that, please have a look in detail again :)